company.el

```diff
@@ -645,18 +645,48 @@ commands in the `company-' namespace, abort completion."
                         (repeat :tag "Commands" function))
                  (repeat :tag "Commands" function)))
 
-(defcustom company-show-numbers nil
-  "If enabled, show quick-access numbers for the first ten candidates."
+(define-obsolete-variable-alias
+ 'company-show-numbers
+ 'company-show-quick-access
+ "0.9.14")
+
+(defcustom company-show-quick-access nil
+  "Show quick-access keys beside the candidates.
+
+Non-nil value enables a quick-access column on the right side of the tooltip,
+unless the configured value is a symbol `left'.
+If the value is nil, do not show a quick-access column.
+
+To change the quick-access keys shown, customize `company-quick-access-keys'."
   :type '(choice (const :tag "off" nil)
                  (const :tag "left" 'left)
                  (const :tag "on" 't)))
 
-(defcustom company-show-numbers-function #'company--show-numbers
-  "Function called to get quick-access numbers for the first ten candidates.
-
-The function receives the candidate number (starting from 1) and should
-return a string prefixed with one space."
-  :type 'function)
+(define-obsolete-variable-alias
+ 'company-show-numbers-function
+ 'company-quick-access-keys
+ "0.9.14")
+
+(define-obsolete-function-alias
+ 'company--show-numbers
+ 'company-quick-access-keys
+ "0.9.14")
+
+(defcustom company-quick-access-keys
+  '("1" "2" "3" "4" "5" "6" "7" "8" "9" "0")
+  "A list of quick-access keys.
+Quick access is bound to each of `M-VALUE' key sequence.
+If `company-show-quick-access' is non-nil, a quick-access column
+shows the configured quick-access keys beside candidates."
+  :type '(repeat string)
+  :initialize #'custom-initialize-default
+  :set #'(lambda (sym key)
+           (custom-set-default sym key)
+           (company-active-map-builder)
```

Comment:
It would be better for this setter not to recompute the whole keymaps, and instead only modify the existing value of the keymap. Because otherwise we can have a user who has changed the keys in, say, `company-active-map`, then they change `company-quick-access-keys` through Customize - and boom, all their previous changes are gone.

This way you won't need to call `company-enable-overriding-keymap` either.

How to do that:

- Have a function like `company--add-quick-access-keys (keymap)`, and call it from both keymap definitions.
- This setter would call it on both keymaps, if those variables are `boundp` (which they won't be the first time this defcustom is loading).
- Also have a function `company--remove-quick-access-keys (keymap)`, which would remove the previous bindings. You would call it before `custom-set-default`, when `company-quick-access-keys` is still set to the previous value (or to no value at all). A `boundp` check is probably necessary there as well.

Reply:
Alternatively, we could avoid this whole setter business and say that changing this variable won't have any effect if the package is already loaded (there are examples of such variables in Emacs as well, IIRC). And suggest, perhaps, to instead call some new function which would perform the necessary modification.

The function could take the values of the new options as argument: the modifier, and the list of keys. It would look at the current value of the var to see which keys to remove, and then change the values in the end.

This might be easier to implement, though the other approach is probably a little more user-friendly.